File: applivery-updates/src/main/java/com/applivery/updates/DownloadService.java
package com.applivery.updates;

import android.app.IntentService;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.Environment;
import android.widget.Toast;

import com.applivery.updates.data.DownloadApiService;
import com.applivery.updates.data.DownloadApiServiceBuilder;
import com.applivery.updates.data.UpdatesApiService;
import com.applivery.updates.domain.Download;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.BufferedInputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

import androidx.core.app.NotificationCompat;
import kotlin.coroutines.CoroutineContext;
import kotlin.coroutines.jvm.internal.BaseContinuationImpl;
import okhttp3.ResponseBody;
import retrofit2.Call;

public class DownloadService extends IntentService {

    public DownloadService() {
        super("Download Service");
    }

    private static final String CHANNEL_ID = "CHANNEL_ID";

    private NotificationCompat.Builder notificationBuilder;
    private NotificationManager notificationManager;
    private int totalFileSize;
    private DownloadApiServiceBuilder downloadApiServiceBuilder;

    @Override
    protected void onHandleIntent(Intent intent) {
        createNotificationChannel();
        notificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);

        notificationBuilder = new NotificationCompat.Builder(this, CHANNEL_ID)
                .setSmallIcon(R.drawable.ic_download)
                .setContentTitle("Download")
                .setContentText("Downloading File")
                .setAutoCancel(true);
        notificationManager.notify(0, notificationBuilder.build());

        initDownload();
    }

    private void initDownload() {

        downloadApiServiceBuilder = new DownloadApiServiceBuilder();

        DownloadApiService downloadApiService = downloadApiServiceBuilder.build();
        Call<ResponseBody> request = downloadApiService.downloadBuild("");
        try {
            downloadFile(request.execute().body());
        } catch (IOException e) {
            e.printStackTrace();
            Toast.makeText(getApplicationContext(), e.getMessage(), Toast.LENGTH_SHORT).show();
        }
    }

    private void downloadFile(ResponseBody body) throws IOException {
        int count;
        byte data[] = new byte[1024 * 4];
        long fileSize = body.contentLength();
        InputStream bis = new BufferedInputStream(body.byteStream(), 1024 * 8);
        File outputFile = new File(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS), "file.zip");
        OutputStream output = new FileOutputStream(outputFile);
        long total = 0;
        long startTime = System.currentTimeMillis();
        int timeCount = 1;
        while ((count = bis.read(data)) != -1) {

            total += count;
            totalFileSize = (int) (fileSize / (Math.pow(1024, 2)));
            double current = Math.round(total / (Math.pow(1024, 2)));

            int progress = (int) ((total * 100) / fileSize);

            long currentTime = System.currentTimeMillis() - startTime;

            Download download = new Download();
            download.setTotalFileSize(totalFileSize);

            if (currentTime > 1000 * timeCount) {

                download.setCurrentFileSize((int) current);
                download.setProgress(progress);
                sendNotification(download);
                timeCount++;
            }

            output.write(data, 0, count);
        }
        onDownloadComplete();
        output.flush();
        output.close();
        bis.close();

    }

    private void sendNotification(Download download) {
        sendIntent(download);
        notificationBuilder.setProgress(100, download.getProgress(), false);
        notificationBuilder.setContentText(String.format("Downloaded (%d/%d) MB", download.getCurrentFileSize(), download.getTotalFileSize()));
        notificationManager.notify(0, notificationBuilder.build());
    }

    private void sendIntent(Download download) {

//        Intent intent = new Intent(MainActivity.MESSAGE_PROGRESS);
//        intent.putExtra("download",download);
//        LocalBroadcastManager.getInstance(DownloadService.this).sendBroadcast(intent);
    }

    private void onDownloadComplete() {

        Download download = new Download();
        download.setProgress(100);
        sendIntent(download);

        notificationManager.cancel(0);
        notificationBuilder.setProgress(0, 0, false);
        notificationBuilder.setContentText("File Downloaded");
        notificationManager.notify(0, notificationBuilder.build());

    }

    @Override
    public void onTaskRemoved(Intent rootIntent) {
        notificationManager.cancel(0);
    }

    private void createNotificationChannel() {
        // Create the NotificationChannel, but only on API 26+ because
        // the NotificationChannel class is new and not in the support library
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            CharSequence name = getString(R.string.channel_name);
            String description = getString(R.string.channel_description);
            int importance = NotificationManager.IMPORTANCE_DEFAULT;
            NotificationChannel channel = new NotificationChannel(CHANNEL_ID, name, importance);
            channel.setDescription(description);
            // Register the channel with the system; you can't change the importance
            // or other notification behaviors after this
            NotificationManager notificationManager = getSystemService(NotificationManager.class);
            notificationManager.createNotificationChannel(channel);
        }
    }
}

File: applivery-updates/src/main/java/com/applivery/updates/DownloadService.kt
package com.applivery.updates

import android.app.IntentService
import android.app.NotificationChannel
import android.app.NotificationManager
import android.content.Context
import android.content.Intent
import android.os.Build
import androidx.core.app.NotificationCompat
import com.applivery.base.AppliveryDataManager
import com.applivery.base.util.AppliveryLog
import com.applivery.updates.data.ApiServiceProvider
import com.applivery.updates.data.DownloadApiService
import com.applivery.updates.data.UpdatesApiService
import com.applivery.updates.domain.DownloadInfo
import com.applivery.updates.util.ApkInstaller
import okhttp3.ResponseBody
import java.io.File
import java.io.IOException


private const val NOTIFICATION_CHANNEL_ID = "NOTIFICATION_CHANNEL_87234"
private const val NOTIFICATION_ID = 0x21

class DownloadService : IntentService("Download apk service") {

    private var notificationBuilder: NotificationCompat.Builder? = null
    private var notificationManager: NotificationManager? = null
    private lateinit var fileManager: FileManager
    private lateinit var downloadApiService: DownloadApiService
    private lateinit var updatesApiService: UpdatesApiService

    override fun onHandleIntent(intent: Intent?) {
        AppliveryDataManager.appData?.run {

            isDownloadStarted = true
            fileManager = FileManager(applicationContext)
            updatesApiService = ApiServiceProvider.getApiService()
            downloadApiService = ApiServiceProvider.getDownloadApiService()

            val apkFileName = name.replace(" ", "-") +
                    "-" + appConfig.lastBuildId + ".apk"

            // TODO check if file exits

            val buildToken = getBuildToken(appConfig.lastBuildId)
            if (buildToken.isNotEmpty()) {
                initDownload(name, apkFileName, buildToken)
            }
        } ?: also {
            isDownloadStarted = false
            AppliveryLog.error("The download cannot be started with null app data")
        }
    }

    private fun getBuildToken(buildId: String): String {
        return try {
            val tokenResponse = updatesApiService.obtainBuildToken(buildId).execute()

            if (tokenResponse.isSuccessful) {
                tokenResponse.body()?.data?.token as String
            } else {
                // TODO parse error to show the error
                AppliveryLog.error("Invalid config. Cannot get the build token")
                ""
            }
        } catch (e: Exception) {
            AppliveryLog.error("Cannot get the build token")
            ""
        }
    }

    private fun initDownload(appName: String, apkFileName: String, buildToken: String) {
        showNotification(appName)
        try {
            val request = downloadApiService.downloadBuild(buildToken)
            request.execute().body()?.run { downloadFile(apkFileName, this) }
        } catch (e: IOException) {
            notificationManager?.cancel(0)
            AppliveryLog.error("Error downloading apk")
        }
    }

    @Throws(IOException::class)
    private fun downloadFile(apkFileName: String, body: ResponseBody) {

        val outputFile = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            File(cacheDir, apkFileName)
        } else {
            File(filesDir, apkFileName)
        }

        fileManager.writeResponseBodyToDisk(body = body,
            file = outputFile,
            onUpdate = { downloadInfo ->
                updateProgress(downloadInfo)
                ProgressListener.downloadInfo = downloadInfo
            },
            onFinish = {
                ProgressListener.onFinish?.invoke()
                isDownloadStarted = false
                onDownloadComplete(outputFile.path)
            },
            onError = {
                isDownloadStarted = false
                AppliveryLog.error("Error saving apk")
            })
    }

    private fun showNotification(appName: String) {
        createNotificationChannel()
        notificationManager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

        notificationBuilder = NotificationCompat.Builder(this, NOTIFICATION_CHANNEL_ID)
            .setSmallIcon(R.drawable.ic_applivery_blue)
            .setContentTitle(getString(R.string.applivery_updates_app_name))
            .setContentText(getString(R.string.applivery_updates_notification_text, appName))
            .setAutoCancel(false)
            .setOngoing(true)
            .setProgress(100, 0, true)
        notificationManager?.notify(NOTIFICATION_ID, notificationBuilder?.build())
    }

    private fun updateProgress(downloadInfo: DownloadInfo) {
        notificationBuilder?.setProgress(100, downloadInfo.progress, false)
        notificationBuilder?.setContentText(
            getString(
                R.string.applivery_updates_notification_progress,
                downloadInfo.currentFileSize,
                downloadInfo.totalFileSize
            )
        )
        notificationManager?.notify(NOTIFICATION_ID, notificationBuilder!!.build())


    }

    private fun onDownloadComplete(filePath: String) {
        clearNotification()
        ApkInstaller.installApplication(this, filePath)
    }

    override fun onTaskRemoved(rootIntent: Intent) {
        clearNotification()
    }

    private fun clearNotification() {
        notificationManager?.cancel(NOTIFICATION_ID)
    }

    private fun createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val name = getString(R.string.applivery_updates_channel_name)
            val description = getString(R.string.applivery_updates_channel_description)
            val importance = NotificationManager.IMPORTANCE_DEFAULT
            val channel = NotificationChannel(NOTIFICATION_CHANNEL_ID, name, importance)
            channel.description = description
            val notificationManager = getSystemService(NotificationManager::class.java)
            notificationManager!!.createNotificationChannel(channel)
        }
    }

    companion object {
        private var isDownloadStarted = false

        fun startDownloadService(context: Context) {
            if (!isDownloadStarted) {
                val intent = Intent(context, DownloadService::class.java)
                context.startService(intent)
            }
        }
    }
}